src/app/icd/client/DefaultICDClientStorage.cpp

```diff
@@ -0,0 +1,397 @@
+
+    TLV::TLVType arrayType;
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Array, arrayType));
+    for (auto & fabric : mFabricList)
+    {
+        ReturnErrorOnFailure(writer.Put(TLV::AnonymousTag(), fabric));
+    }
+    return writer.EndContainer(arrayType);
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    writer.Finalize(backingBuffer);
+    ReturnErrorOnFailure(mpClientInfoStore->SyncSetKeyValue(DefaultStorageKeyAllocator::ICDFabricList().KeyName(),
+                                                                     backingBuffer.Get(), static_cast<uint16_t>(len)));
+    uint16_t counterLen = static_cast<uint16_t>(sizeof(counter));
+    ReturnErrorOnFailure(mpClientInfoStore->SyncDeleteKeyValue(
+        DefaultStorageKeyAllocator::ICDFabricListCounter().KeyName()));
+    return mpClientInfoStore->SyncSetKeyValue(
+        DefaultStorageKeyAllocator::ICDFabricListCounter().KeyName(), &counter, counterLen);
+}
+
+CHIP_ERROR DefaultICDClientStorage::LoadFabricList()
+{
+    size_t counter      = 0;
+    uint16_t counterLen = static_cast<uint16_t>(sizeof(counter));
+    ReturnErrorOnFailure(mpClientInfoStore->SyncGetKeyValue(
+        DefaultStorageKeyAllocator::ICDFabricListCounter().KeyName(), &counter, counterLen));
+
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    size_t len = FABRIC_INDEX_TLV_SIZE * counter;
+    ReturnErrorCodeIf(!backingBuffer.Calloc(len), CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+    uint16_t length = static_cast<uint16_t>(len);
+    ReturnErrorOnFailure(mpClientInfoStore->SyncGetKeyValue(DefaultStorageKeyAllocator::ICDFabricList().KeyName(),
+                                                                    backingBuffer.Get(), length));
+
+    TLV::ScopedBufferTLVReader reader(std::move(backingBuffer), len);
+    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Array, TLV::AnonymousTag()));
+    TLV::TLVType arrayType;
+    ReturnErrorOnFailure(reader.EnterContainer(arrayType));
+
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    while ((err = reader.Next(TLV::kTLVType_UnsignedInteger, TLV::AnonymousTag())) == CHIP_NO_ERROR)
+    {
+        FabricIndex fabricIndex;
+        ReturnErrorOnFailure(reader.Get(fabricIndex));
+        mFabricList.push_back(fabricIndex);
+    }
+
+    if (err != CHIP_END_OF_TLV)
+    {
+        return err;
+    }
+    ReturnErrorOnFailure(reader.ExitContainer(arrayType));
+    return reader.VerifyEndOfContainer();
+}
+
+DefaultICDClientStorage::ICDClientInfoIteratorImpl::ICDClientInfoIteratorImpl(DefaultICDClientStorage & manager) : mManager(manager)
+{
+    mFabricListIndex    = 0;
+    mClientInfoIndex = 0;
+    mClientInfoVector.clear();
+}
+
+size_t DefaultICDClientStorage::ICDClientInfoIteratorImpl::Count()
+{
+    size_t total = 0;
+    for (auto & fabric : mManager.mFabricList)
```

Comment:
All the `auto & fabric` should be `auto & fabric_idx`

Reply:
updated, thanks